### PR TITLE
Arm64: Simplify TryEncodeBitMask and use for constants

### PR DESCRIPTION
**Pull request by merryhime**
* O(1) algorithm for determining bit immediates.
* Detect when bit immediates are usable for materalizing constants.
* Very minor improvement in performance on arm64.

### CODE REVIEW

File: ARMeilleure/CodeGen/Arm64/CodeGenCommon.cs [hunk 1]
```diff
@@ -140,34 +78,11 @@ public static bool TryEncodeBitMask(ulong value, out int immN, out int immS, out
...
-            }
-            else
-            {
-                immN = 0;
-                immS = (0xf80 >> BitOperations.Log2((uint)repetitions)) & 0x3f;
-            }
-
-            immS |= onesCount - 1;
+            immN = (elementSize >> 6) & 1;
+            immS = (((~elementSize + 1) << 1) | (onesInElement - 1)) & 0x3f;
```
Comment (merryhime) [inline ARMeilleure/CodeGen/Arm64/CodeGenCommon.cs:82]: I'm happy to switch this to `(0xf80 >> BitOperations.Log2((uint)repetitions))` if that makes more sense to you.

**Comment by gdkchan** [comment]
Might be worth changing `CodeGenerator.GenerateCopy` to also try using the bitmask mov if possible, since that's just one instruction. That was not done before because the `TryEncodeBitMask` function was pretty expensive, so  checking it for every copy could be a bit costly.

File: ARMeilleure/CodeGen/Arm64/CodeGenCommon.cs (hunk 1 above)
Comment (gdkchan) [reply ARMeilleure/CodeGen/Arm64/CodeGenCommon.cs:82]: `(~elementSize + 1)` could be `-elementSize` I guess?

**Comment by merryhime** [comment]
> Might be worth changing CodeGenerator.GenerateCopy to also try using the bitmask mov if possible, since that's just one instruction.

Done I think.

**Review by gdkchan — APPROVED** [review]
lgtm, thanks. Tested a bunch of games on my m1 macbook and they are working fine.

**Review by marysaka — APPROVED** [review]

**Pull request merged**